modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +330,168 @@ void run(ClusterState state) {
+                    transportActionsDeduplicator.executeOnce(
+                        addIndexBlockRequest,
+                        new ErrorRecordingActionListener(indexName, errorStore),
+                        (req, reqListener) -> addIndexBlock(addIndexBlockRequest, reqListener)
+                    );
+                    continue;
+                }
+
+                if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource)
+                    && backingIndexDownsamplingStatus.equals(SUCCESS)) {
+                    // if the backing index is a downsample index itself, let's check if its source index still exists as we must delete it
+                    Map<String, String> lifecycleMetadata = backingIndex.getCustomData(LIFECYCLE_CUSTOM_INDEX_METADATA_KEY);
+                    if (lifecycleMetadata == null || lifecycleMetadata.containsKey(REPLACEMENT_SOURCE_INDEX) == false) {
+                        // this is a downsampling index that was not added by data stream lifecycle in this data stream
+                        continue;
+                    }
+                    String actualDownsamplingSource = lifecycleMetadata.get(REPLACEMENT_SOURCE_INDEX);
+                    IndexMetadata downsampleSourceIndex = metadata.index(actualDownsamplingSource);
+                    if (downsampleSourceIndex != null) {
+                        affectedIndices.add(downsampleSourceIndex.getIndex());
+                        // delete downsampling source index (that's not part of the data stream anymore) before doing any more
+                        // downsampling
+                        DeleteIndexRequest deleteIndexRequest = new DeleteIndexRequest(backingIndexDownsamplingSource).masterNodeTimeout(
+                            TimeValue.MAX_VALUE
+                        );
+                        transportActionsDeduplicator.executeOnce(
+                            deleteIndexRequest,
+                            new ErrorRecordingActionListener(backingIndexDownsamplingSource, errorStore),
+                            (req, reqListener) -> deleteIndex(
+                                deleteIndexRequest,
+                                "replacement with its downsampled index in the data stream",
+                                reqListener
+                            )
+                        );
+                        continue;
+                    }
+                }
+
+                // this index has matching downsample rounds and is read-only, so let's wait for an in-progress downsampling operation to
+                // succeed or trigger the corresponding one
+                // TODO add support for cancelling a current in-progress operation if another, later, round matches
+                DataStreamLifecycle.Downsampling.Round lastRound = downsamplingRounds.get(downsamplingRounds.size() - 1);
+
+                for (DataStreamLifecycle.Downsampling.Round round : downsamplingRounds) {
```

Comment:
As I mentioned in a previous comment, I think we should factor this out into a separate method so it's easier to digest (it's still huge even as a single method).

Reply:
Great shout. This is now done